aten/src/ATen/native/cpu/HistogramKernel.cpp

```diff
@@ -0,0 +1,179 @@
+                    int64_t pos_min = std::max(static_cast<int64_t>(0), pos - 1);
+                    int64_t pos_max = std::min(pos + 2, numel_be);
+                    pos = std::upper_bound(data_be + pos_min, data_be + pos_max, elt) - data_be - 1;
+                }
+            } else {
+                TORCH_INTERNAL_ASSERT(false);
+            }
+
+            // Unlike other bins, the rightmost bin includes its right boundary
+            if (pos == (numel_be - 1)) {
+                pos -= 1;
+            }
+
+            // In the unweighted case, the default weight is 1
+            input_t wt = accessor_wt.has_value() ? accessor_wt.value()[i] : static_cast<input_t>(1);
+            data_out_local[pos] += wt;
+        }
+
+        // Locks and updates the common output
+        const std::lock_guard<std::mutex> lock(data_out_mutex);
+        for (int64_t i = 0; i < numel_be - 1; i++) {
+            data_out[i] += data_out_local[i];
+        }
+    });
+}
+
+/* Some pre- and post- processing steps for the main algorithm.
+ * Initializes hist to 0, calls into the main algorithm, and normalizes output if necessary.
+ */
+template<BIN_SELECTION_ALGORITHM bin_algorithm>
+void histogram_out_cpu_template(const Tensor& self, const c10::optional<Tensor>& weight, bool density,
+        Tensor& hist, const Tensor& bin_edges) {
+    hist.fill_(0);
+
+    const int64_t numel_in = self.numel();
+    const Tensor reshaped_input = self.reshape({numel_in});
+
+    const auto reshaped_weight = weight.has_value()
+            ? c10::optional<Tensor>(weight.value().reshape({numel_in}))
+            : c10::optional<Tensor>();
+
+    switch (self.scalar_type()) {
+        case ScalarType::Double: {
+            histogram_cpu_contiguous<double, bin_algorithm>(
+                  hist, bin_edges.contiguous(), reshaped_input, reshaped_weight);
+            break;
+        }
+        case ScalarType::Float: {
+            histogram_cpu_contiguous<float, bin_algorithm>(
+                  hist, bin_edges.contiguous(), reshaped_input, reshaped_weight);
+            break;
+        }
+        default:
+            TORCH_CHECK(false, "torch.histogram: not supported on CPU for dtype ", self.dtype());
+    }
+
+    if (density) {
+        auto bin_widths = bin_edges.diff();
+        auto hist_sum = hist.sum().item();
+        // Converts the bin totals to a probability density function by dividing by the bin widths
+        hist.div_(bin_widths);
+        // Normalizes the PDF so that the integral over the bins is 1
+        hist.div_(hist_sum);
+    }
+}
+
+static void histogram_kernel_impl(const Tensor& self, const c10::optional<Tensor>& weight, bool density,
+        Tensor& hist, const Tensor& bin_edges) {
+    histogram_out_cpu_template<BINARY_SEARCH>(self, weight, density, hist, bin_edges);
+}
+
+static void histogram_linear_kernel_impl(const Tensor& self, const c10::optional<Tensor>& weight,
```

Comment:
What's the performance of the linear interpolated search vs the bin search, and why have "local search" for histogram but not for histc?

Reply:
In the case of **histogram**, the output is a tuple of (histogram, bin_edges). bin_edges are constructed by linspace and may not perfectly align with bucketing of the input values by the `LINEAR_INTERPOLATION` approach. `LINEAR_INTERPOLATION_WITH_LOCAL_SEARCH` ensures that the output is consistent within itself; the returned histogram will match what would be expected when bucketing the values in the original input according to the returned bin_edges.

In the case of **histc**, the output is only the histogram tensor. That means we can bucket by `LINEAR_INTERPOLATION` without worrying about producing a sequence of `bin_edges` consistent with the resulting histogram. Additionally, the `LINEAR_INTERPOLATION` behavior matches the pre-existing implementation of histc.

I need to redo the full benchmarks after all the recent changes, but from initial testing the "local search" seems to add about 50% overhead when compared to pure linear interpolation.

Reply:
That's a cool analysis -- would you add a comment explaining it and the measured perf impact to the code?

How much faster is linear interpolation (with search) than bin searching?

Reply:
Summary of local benchmarking results:

![image](https://user-images.githubusercontent.com/3241341/121595477-50443e00-ca0c-11eb-9a54-362d8fb6b46a.png)